library/ceph_volume.py

```diff
@@ -250,6 +240,146 @@ def run_module():
     module.exit_json(**result)
 
 
+def remove_osd(module):
+    """
+    If the OSD exists, it will 'ceph osd destroy' the
+    OSD. If zap is set to True then the devices, lvs and
+    partitions used to create the OSD will be zapped with
+    'ceph-volume lvm zap' even if the OSD using the data
+    device is not still active.
+    """
+    data = module.params['data']
+    data_vg = module.params.get('data_vg', None)
+    journal = module.params.get('journal', None)
+    journal_vg = module.params.get('journal_vg', None)
+    db = module.params.get('db', None)
+    db_vg = module.params.get('db_vg', None)
+    wal = module.params.get('wal', None)
+    wal_vg = module.params.get('wal_vg', None)
+    zap = module.params.get('zap', False)
+
+    base_zap_cmd = [
+        'ceph-volume',
+        'lvm',
+        'zap',
+        # for simplicity always --destroy. It will be needed
+        # for raw devices and will noop for lvs.
+        '--destroy',
+    ]
+
+    commands = []
+
+    data = get_data(data, data_vg)
+
+    # check to see if osd already exists
+    # FIXME: this does not work when data is a raw device
+    # support for 'lvm list' and raw devices was added with https://github.com/ceph/ceph/pull/20620 but
+    # has not made it to a luminous release as of 12.2.4
+    rc, out, err = module.run_command(["ceph-volume", "lvm", "list", "--format=json", data], encoding=None)
```

Comment:
this is mixing two things that might not work well in all situations:

* `lvm list` can list OSDs that are present in the node, but are absent from the cluster
* `osd destroy` will tell the cluster that the OSD ID has to be removed (keeping the ID around for re-use) but will fail if the ID is no longer active in the cluster

In this case it would need to do something similar to what ceph-volume does, which is check the cluster:

    bootstrap_keyring = '/var/lib/ceph/bootstrap-osd/%s.keyring' % conf.cluster
    stdout, stderr, returncode = process.call(
        [
            'ceph',
            '--cluster', conf.cluster,
            '--name', 'client.bootstrap-osd',
            '--keyring', bootstrap_keyring,
            'osd',
            'tree',
            '-f', 'json',
        ],
        show_command=True
    )
    if returncode != 0:
        raise RuntimeError('Unable check if OSD id exists: %s' % osd_id)

    output = json.loads(''.join(stdout).strip())
    osds = output['nodes']
    return any([str(osd['id']) == str(osd_id) for osd in osds])

Since `state=absent` is trying make sure the OSD ID is removed from both the node and the cluster, either error checking the return of `destroy_cmd` or checking with the cluster first would be required.